### PR TITLE
 Add tool elasticsearch-node unsafe-bootstrap

### PR DESCRIPTION
**Pull request by andrershov**
This tool could be used if the majority of master-eligible nodes is lost.
Docs can be found in https://github.com/elastic/elasticsearch/pull/37812

### CODE REVIEW

**Comment by andrershov** [comment]
run grader build tests 2

**Comment by andrershov** [comment]
run gradle build tests 2

**Comment by andrershov** [comment]
run elasticsearch-ci 2

**Comment by andrershov** [comment]
run elasticsearch-ci-2

**Comment by andrershov** [comment]
run elasticsearch-ci/2

**Comment by andrershov** [comment]
run elasticsearch-ci/2

File: server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+public class UnsafeBootstrapMasterCommand extends EnvironmentAwareCommand {
+
+    private final NamedXContentRegistry namedXContentRegistry;
+
+    static final String WARNING_MSG =
+            "-----------------------------------------------------------------------\n" +
+            "\n" +
+            "    WARNING: Elasticsearch MUST be stopped before running this tool.\n" +
+            "\n" +
+            "You should run this tool only if you've lost the majority of master eligible nodes.\n" +
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java]: Suggest:

> You should run this tool only if you have permanently lost half or more of the master-eligible nodes, and you cannot restore the cluster from a snapshot. This tool can result in arbitrary data loss and should be the last resort.

I'm suggesting to avoid saying "lost a majority" because you also need this tool if you have an even number of nodes and you've lost half of them, which isn't strictly a majority.

Also could you explain how it can render the cluster "completely non-functional"? I mean I see how it can break indices, but it should allow a cluster to form and any broken indices can be deleted.

File: server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    static final String NO_GLOBAL_METADATA_MSG = "failed to find global metadata, metadata corrupted?";
+    static final String EMPTY_LAST_COMMITTED_VOTING_CONFIG_MSG =
+            "last committed voting voting configuration is empty, cluster has never been bootstrapped?";
+    static final String WRITE_METADATA_EXCEPTION_MSG = "exception occurred when writing new metadata to disk";
+    static final String MASTER_NODE_BOOTSTRAPPED_MSG = "Master node was successfully bootstrapped";
+    static final Setting<String> UNSAFE_BOOTSTRAP =
+            ClusterService.USER_DEFINED_META_DATA.getConcreteSetting("cluster.metadata.unsafe-bootstrap");
+
+    UnsafeBootstrapMasterCommand() {
+        super("Unsafely bootstraps the master node if the majority of master eligible nodes is lost");
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java]: Again, not strictly a majority. Maybe we should replace "bootstraps" with "forces the successful election of"?

File: server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+            throw new ElasticsearchException(ABORTED_BY_USER_MSG);
+        }
+
+        Settings settings = env.settings();
+        terminal.println(Terminal.Verbosity.VERBOSE, "Checking node.master setting");
+        Boolean master = Node.NODE_MASTER_SETTING.get(settings);
+        if (master == false) {
+            throw new ElasticsearchException(NOT_MASTER_NODE_MSG);
+        }
+        final int nodeOrdinal = 0;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java]: I think that we need to be able to deal with higher node ordinals somehow. To be discussed, as it's not immediately clear how to do so.

File: server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+                .coordinationMetaData(newCoordinationMetaData)
+                .build();
+        writeNewMetaData(terminal, manifest, newMetaData, dataPaths);
+    }
+
+    private void writeNewMetaData(Terminal terminal, Manifest manifest, MetaData newMetaData, Path[] dataPaths) {
+        try {
+            terminal.println(Terminal.Verbosity.VERBOSE, "Writing new global metadata to disk");
+            long newGeneration = MetaData.FORMAT.write(newMetaData, dataPaths);
+            Manifest newManifest = new Manifest(manifest.getCurrentTerm(), manifest.getClusterStateVersion(), newGeneration,
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java]: I think we should increment the term, and log that we've done so.

File: server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+    UnsafeBootstrapMasterCommand() {
+        super("Unsafely bootstraps the master node if the majority of master eligible nodes is lost");
+        namedXContentRegistry = new NamedXContentRegistry(ClusterModule.getNamedXWriteables());
+    }
+
+    @Override
+    protected void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
+        // We don't use classical "private static final Logger", because log4j2 initialization happens
+        // after UnsafeBootstrapMasterCommand instance creation.
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java]: Hmm, we make a `private static final Logger logger` in `RemoveCorruptedShardDataCommand`, does that not work?

Also, does this logger have the same configuration as loggers in Elasticsearch proper, i.e., it writes to the Elasticsearch log by default? If so, I think we should log more information about this tool having been run in that log.

File: server/src/test/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterIT.java [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+        }
+
+        expectThrows(() -> executeCommand(environment), UnsafeBootstrapMasterCommand.NO_NODE_METADATA_FOUND_MSG);
+    }
+
+    public void testNotBootstrappedCluster() throws Exception {
+        internalCluster().startNode(
+                Settings.builder()
+                        .put(ElectMasterService.DISCOVERY_ZEN_MINIMUM_MASTER_NODES_SETTING.getKey(), Integer.MAX_VALUE)
+                        .put(DiscoverySettings.INITIAL_STATE_TIMEOUT_SETTING.getKey(), "2s") //to ensure quick node startup
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterIT.java]: Why not zero?

File: server/src/test/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterIT.java [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+import java.util.Collections;
+import java.util.List;
+
+import static org.hamcrest.Matchers.containsString;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0, autoMinMasterNodes = false)
+@TestLogging("_root:DEBUG,org.elasticsearch.cluster.service:TRACE,org.elasticsearch.discovery.zen:TRACE")
+public class UnsafeBootstrapMasterIT extends ESIntegTestCase {
+
+    private int bootstrapNodeId;
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterIT.java]: This seems to be used more as a flag indicating whether or not to bootstrap the cluster, and a default of `do bootstrap the cluster` makes more sense to me.

File: server/src/test/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterIT.java [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+        if (internalCluster().size() + allNodesSettings.size() == bootstrapNodeId) {
+            List<String> nodeNames = new ArrayList<>();
+            Collections.addAll(nodeNames, internalCluster().getNodeNames());
+            allNodesSettings.forEach(settings -> nodeNames.add(Node.NODE_NAME_SETTING.get(settings)));
+
+            List<Settings> otherNodesSettings = allNodesSettings.subList(0, allNodesSettings.size() - 1);
+            Settings lastNodeSettings = allNodesSettings.get(allNodesSettings.size() - 1);
+            List<Settings> newSettings = new ArrayList<>();
+            newSettings.addAll(otherNodesSettings);
+            newSettings.add(Settings.builder().put(lastNodeSettings)
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterIT.java]: Could we bootstrap a random node rather than it always being the last of them?

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
I left mostly minor suggestions, but we should discuss how to deal with node ordinals >0 as I don't think we should just silently ignore them.

I will look in more depth at the docs later.

File: qa/vagrant/src/main/java/org/elasticsearch/packaging/test/ArchiveTestCase.java [hunk 9]
```diff
@@ -318,4 +318,22 @@ public void test100RepairIndexCliPackaging() {
         }
     }
 
+    public void test110RepairIndexCliPackaging() {
```
[the remark this answers is not in the document]
Comment (andrershov) [reply qa/vagrant/src/main/java/org/elasticsearch/packaging/test/ArchiveTestCase.java]: Oops, I've changed the names of both tests in this commit https://github.com/elastic/elasticsearch/pull/37696/commits/e2abf3fce9fd36d83ca0060690e985a50460ea45

File: server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java (hunk 2 above)
Comment (andrershov) [reply server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java]: How do you find this message https://github.com/elastic/elasticsearch/pull/37696/commits/b6ada540cf2d2245bbff85c8d4553b5acfb28011?

File: server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java (hunk 4 above)
Comment (andrershov) [reply server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java]: I've implemented increment in this commit https://github.com/elastic/elasticsearch/pull/37696/commits/0e5091330717968d54887d3f6db7c37b1c3236cc. For now, I'm logging new `currentTerm` to the terminal, will get back to it - once I figure out how our command line tools work with elasticsearch logs.

File: server/src/test/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterIT.java (hunk 6 above)
Comment (andrershov) [reply server/src/test/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterIT.java]: https://github.com/elastic/elasticsearch/pull/37696/commits/6552d541f5eb6bacee120200b4835a2c1cd4da4d

File: server/src/test/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterIT.java (hunk 7 above)
Comment (andrershov) [reply server/src/test/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterIT.java]: Well, in fact, it's not a flag. It's the sequence number of the started node, which is used to bootstrap the cluster. I'm using `autoMinMasterNodes=false`, which means I don't want auto-bootstrap. If we set `bootstrapNodeId` to something different than -1, what value to choose? Shall we bootstrap the cluster after the first node is started or the third node is started? So by default, the behaviour is the same as `autoMinMasterNodes=false`, but the tests that need the cluster bootstrap and understand when this bootstrap should be performed can set this field to some positive value.
Please note, this code is copy pasted in several test classes already, I plan to have a follow-up PR that refactors this method to `ESIntegTestCase` so that tests with `autoMinMasterNodes=false` can have an easy way to perform the bootstrap.

File: server/src/test/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterIT.java (hunk 8 above)
Comment (andrershov) [reply server/src/test/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterIT.java]: https://github.com/elastic/elasticsearch/pull/37696/commits/f3d78c41dd0e9a155166ed21d34733b88a2c45a2

File: server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java (hunk 5 above)
Comment (andrershov) [reply server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java]: Ok, I have figured out how the logging works:

1. Most of the commands, for example, `LoggingAwareCommand` are passing `CommandLoggingConfigurator::configureLoggingWithoutConfig` callback to `Command` constructor as `beforeMain` argument. This is the function that will be executed when `Command.main` is called.
2. From CommandLoggingConfigurator::configureLoggingWithoutConfig javadoc `Configure logging without reading a log4j2.properties file, effectively configuring the status logger and all loggers to the console.` Basically it means that logs won't be appended to elasticsearch log files.
3. If you run `ShardToolCli` indeed you get `ERROR StatusLogger No Log4j 2 configuration file found. Using default configuration`. The reason for this is because logger initialization happens when log4j2 classes are loaded. Java Runtime loads classes when it first encounters this class, in case of `ShardToolCli` it will load `RemoveCorruptedShardDataCommand` and transitively `Logger` class, because there is static reference to `Logger` class in `RemoveCorruptedShardDataCommand`. Since pre-main is not yet executed, logging is not configured.
4. It turned out that `NodeToolCli` still suffers from the same warning, because there is `ClusterService` (which in turn staticly references `Logger`) static reference in `UnsafeBootstrapMasterCommand`. The reason why I've not noticied it earlier, because I've added this setting last moment, after I've checked how binaries are working.
5. For now I've made the following fix https://github.com/elastic/elasticsearch/pull/37696/commits/e23f59af291f0bc60af9b2a7bd252815950632b9
6. I suggest to revisit how logging works for commands in general in the follow-up PR. It seems that `beforeMain` could be completely removed and replaced by logging initialization in the command constructor. But this change touches a lot of file and I prefer separate PR.

**Comment by andrershov** [comment]
@DaveCTurner thanks for your review! I've addressed all the comments except `nodeOrdinal`, I think it's ready for the 2nd review pass.

File: server/src/test/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterIT.java (hunk 7 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterIT.java]: Ok

File: server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java]: We discussed and concluded there is no good way to do this, but also no real need to support higher node ordinals.

File: docs/reference/commands/node-tool.asciidoc [hunk 10]
```diff
@@ -0,0 +1,119 @@
+[[node-tool]]
+== elasticsearch-node
```
Comment (DaveCTurner) [inline docs/reference/commands/node-tool.asciidoc]: We decided to move the docs changes to a separate PR - #37812, so this should go.

**Review by DaveCTurner — COMMENTED** [review]
Ok, looks good apart from the removal of the docs that we discussed.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by andrershov** [comment]
run elasticsearch-ci/1

**Comment by andrershov** [comment]
run elasticsearch-ci/1

**Review by DaveCTurner — APPROVED** [review]
LGTM again
